### PR TITLE
Add missing Intel hardware intrinsics and fix generic APIs

### PR DESCRIPTION
**Pull request by fiigii**
This PR 
- add missing APIs discussed in https://github.com/dotnet/corefx/issues/25926
- fix generic names
- fix comments
- disambiguate with `using static`

### CODE REVIEW

File: src/mscorlib/src/System/Runtime/Intrinsics/X86/Avx.PlatformNotSupported.cs [hunk 1]
```diff
@@ -183,36 +183,36 @@ public static class Avx
...
         /// <summary>
         /// __int16 _mm256_extract_epi16 (__m256i a, const int index)
         /// </summary>
-        public static ushort ExtractUInt16<T>(Vector256<T> value, byte index) where T : struct { throw new PlatformNotSupportedException(); }
+        public static ushort Extract(Vector256<ushort> value, byte index) { throw new PlatformNotSupportedException(); }
         /// <summary>
         /// __int32 _mm256_extract_epi32 (__m256i a, const int index)
         /// </summary>
-        public static int ExtractInt32<T>(Vector256<T> value, byte index) where T : struct { throw new PlatformNotSupportedException(); }
+        public static int Extract(Vector256 value, byte index) { throw new PlatformNotSupportedException(); }
```
Comment (4creators) [inline src/mscorlib/src/System/Runtime/Intrinsics/X86/Avx.PlatformNotSupported.cs]: nit: is Vector256 should be Vector256<int>

**Comment by fiigii** [comment]
Fixed some comments and I have double checked. Nothing missing now.
cc @tannergooding @jkotas

**Comment by jkotas** [comment]
Could you please prepare PR with matching change in CoreFX? LGTM otherwise.

**Comment by fiigii** [comment]
> A minor request is that you modify the top comment to have a summary of the changes.

> Could you please prepare PR with matching change in CoreFX? 

Will do.

File: src/mscorlib/src/System/Runtime/Intrinsics/X86/Avx.PlatformNotSupported.cs [hunk 2]
```diff
@@ -197,36 +197,36 @@ public static class Avx
         /// <summary>
         /// __int8 _mm256_extract_epi8 (__m256i a, const int index)
         /// </summary>
-        public static sbyte ExtractSByte<T>(Vector256<T> value, byte index) where T : struct { throw new PlatformNotSupportedException(); }
+        public static sbyte Extract(Vector256<sbyte> value, byte index) { throw new PlatformNotSupportedException(); }
```
Comment (tannergooding) [inline src/mscorlib/src/System/Runtime/Intrinsics/X86/Avx.PlatformNotSupported.cs]: @jkotas, you requested many similar functions be made generic in the ARM proposal. Should the same be done here and elsewhere in the API surface?

**Comment by fiigii** [comment]
@tannergooding Why does the CROSS check fail? I have no permission to see the details.

File: src/mscorlib/src/System/Runtime/Intrinsics/X86/Avx.PlatformNotSupported.cs (hunk 2 above)
Comment (4creators) [reply src/mscorlib/src/System/Runtime/Intrinsics/X86/Avx.PlatformNotSupported.cs]: AFAIR The idea was to use generic for functions with parameter types comprising majority or all supported types. In hte case of `Extract` functions there is only limited number of overloads.

File: src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.PlatformNotSupported.cs [hunk 3]
```diff
@@ -269,7 +269,17 @@ public static class Sse
         /// <summary>
         /// __m128 _mm_load_ps (float const* mem_address)
         /// </summary>
-        public static unsafe Vector128<float> LoadAligned(float* address) { throw new PlatformNotSupportedException(); }
+        public static unsafe Vector128<float> LoadAlignedVector128(float* address) { throw new PlatformNotSupportedException(); }
+
+                /// <summary>
+        /// __m128 _mm_loadh_pi (__m128 a, __m64 const* mem_addr)
+        /// </summary>
+        public static unsafe Vector128<float> LoadHigh(Vector128<float> value, float* address) { throw new PlatformNotSupportedException(); }
```
Comment (tannergooding) [inline src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.PlatformNotSupported.cs]: I thought we weren't exposing any of the __m64 intrinsics?

File: src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse41.PlatformNotSupported.cs [hunk 4]
```diff
@@ -407,7 +396,7 @@ public static class Sse41
         /// <summary>
         /// int _mm_test_all_ones (__m128i a)
         /// </summary>
-        public static bool TestAllOnes(Vector128<sbyte> value) { throw new PlatformNotSupportedException(); }
+         public static bool TestAllOnes(Vector128<sbyte> value) { throw new PlatformNotSupportedException(); }
```
Comment (tannergooding) [inline src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse41.PlatformNotSupported.cs]: nit: extra whitespace at beginning of signature

File: src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse41.PlatformNotSupported.cs [hunk 5]
```diff
@@ -451,7 +440,7 @@ public static class Sse41
         public static bool TestMixOnesZeros(Vector128<uint> left, Vector128<uint> right) { throw new PlatformNotSupportedException(); }
         public static bool TestMixOnesZeros(Vector128<long> left, Vector128<long> right) { throw new PlatformNotSupportedException(); }
         public static bool TestMixOnesZeros(Vector128<ulong> left, Vector128<ulong> right) { throw new PlatformNotSupportedException(); }
-        
+
```
Comment (tannergooding) [inline src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse41.PlatformNotSupported.cs]: nit: stray tab on this line

File: src/mscorlib/src/System/Runtime/Intrinsics/X86/Avx.PlatformNotSupported.cs (hunk 2 above)
Comment (jkotas) [reply src/mscorlib/src/System/Runtime/Intrinsics/X86/Avx.PlatformNotSupported.cs]: Right. Here is a link to the discussion for reference: https://github.com/dotnet/corefx/issues/26179#issuecomment-355615654

File: src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.PlatformNotSupported.cs (hunk 3 above)
Comment (4creators) [reply src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.PlatformNotSupported.cs]: It is not MMX intrinsic:
```
__m128 _mm_loadh_pi (__m128 a, __m64 const* mem_addr)
```
It loads to `xmm` register

Comment (fiigii) [reply src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.PlatformNotSupported.cs]: Ah, right.

Comment (tannergooding) [reply src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.PlatformNotSupported.cs]: Looks like its just a confusing signature on the C/C++ side.

The Instruction manual indeed says it only operates on m64 and doesn't deal with _m64 registers at all:
image

Comment (fiigii) [reply src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.PlatformNotSupported.cs]: @4creators Loading data 64-bit by 64-bit into XMM makes no sense on modern CPUs.

Comment (tannergooding) [reply src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.PlatformNotSupported.cs]: (it also doesn't have the "causes a transition to MMX technology" statement)

File: src/mscorlib/src/System/Runtime/Intrinsics/X86/Avx.PlatformNotSupported.cs (hunk 2 above)
Comment (tannergooding) [reply src/mscorlib/src/System/Runtime/Intrinsics/X86/Avx.PlatformNotSupported.cs]: `Extract` is one that operates on 8/10 of the supported `numeric` types, I could considered that a majority.

In either case, it would be good to get a final statement from @jkotas/@CarolEidt so we don't need yet another pr

**Comment by tannergooding** [comment]
> Why does the CROSS check fail? I have no permission to see the details.

@fiigii, looks to be an authentication failure. I re-queued the job to see if it fails again

**Comment by tannergooding** [comment]
Ok, failed with the same. tagging Might need to tag dcenghost if it appears to be impacting other PRs.

File: src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.PlatformNotSupported.cs (hunk 3 above)
Comment (4creators) [reply src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.PlatformNotSupported.cs]: > Loading data 64-bit by 64-bit into XMM is useless on modern CPUs.

In contrary actually. You may want lower 64 bits for one memory location and upper 64 bits from another memory location placed far apart from each other. In such situations this function is a perfect tool and there are many cases where such loads are necessary.

Comment (4creators) [reply src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.PlatformNotSupported.cs]: @fiigii 
As you can see above  it is available even for AVX512 ISA

Comment (tannergooding) [reply src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.PlatformNotSupported.cs]: @4creators, I think the point was that, on modern CPUs there are better instructions for doing this, and even on older CPUs, two loads and a shuffle or Unpack are likely better

Comment (fiigii) [reply src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.PlatformNotSupported.cs]: Okay, let's keep these APIs.

Comment (fiigii) [reply src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.PlatformNotSupported.cs]: We can remove useless APIs later if there is.

Comment (4creators) [reply src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.PlatformNotSupported.cs]: > We can remove useless APIs later if there is.

I think being conservative in deleting APIs which are available in given ISA is the best approach.

Comment (tannergooding) [reply src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.PlatformNotSupported.cs]: @fiigii, if we are keeping it (and we probably should, since it is a regular SSE instruction, and not an MMX Interop instruction), we should also be exposing the `StoreHigh` and `StoreLow` counterparts (for `movhps m64, xmm` and `movlps m64, xmm`)

Comment (tannergooding) [reply src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.PlatformNotSupported.cs]: (and the same for Sse2)

Comment (fiigii) [reply src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.PlatformNotSupported.cs]: > we should also be exposing the StoreHigh and StoreLow counterparts

Thanks, will do.

**Comment by tannergooding** [comment]
👍. Looks like this matches all the feedback so far and covers all the missing intrinsics (at least comparing to the list I compiled before and the lists in Vol 1 of the Intel Architectures Guide.

I would still like to get a definitive answer on when we should be "exploding" the generics and when we should be providing a single generic method.

For example, many of the ones that are still "exploded" are implementing 8/10 or 9/10 of the supported types. Others are only implementing 2/10 or 4/10, for example.

**Comment by fiigii** [comment]
>For example, many of the ones that are still "exploded" are implementing 8/10 or 9/10 of the supported types. Others are only implementing 2/10 or 4/10, for example.

> Right. Here is a link to the discussion for reference: dotnet/corefx#26179 (comment)

@tannergooding @jkotas @CarolEidt 
I just glance over the thread. Seems you guys are talking about something that I tried in the first version of https://github.com/dotnet/corefx/issues/22940. Originally, I folded the class `Sse` into `Sse2` to get beautiful generic APIs and avoid "exploding", but it was rejected. 

**Now, Intel HW intrinsic defines APIs to be generic if and only if the intrinsic covers all the 10/10 numeric types**. I think this rule is good enough (I assume that the "exploding" does not impact JIT throughput), which provides stable runtime behaviors and keeps JIT simple. Meanwhile, with `using static`, non-generic APIs do not hurt user experience than generic ones.

If we want more beautiful generic APIs via moving the bar to
1. _9/10_  that resolves SSE/SSE2 exploding but will be broken by AVX/AVX2.
2. _8/10_ that resolves parts of the AVX/AVX2 exploding but will be broken by someones lacking `long`/`ulong`.
3. _7/10_ that resolves all AVX/AVX2 exploding but will be broken by AVX-512F/CD/BW.
4. ....

Perhaps, we may be able to provide "higher-level" generic intrinsic based-on Intel's staging ISA design. For example, exposing `Vector128<float> Add(...)` in `Sse` and `Vector128<T> Add<T>(...)` in `Sse2`, then `Sse2.Add<float>(..)` generates the SSE instruction. However, that breaks our original design goal of **low-level hardware-specific** intrinsics. Moreover, that will hurt ARM64 intrinsics more.

In summary, I prefer the current rule that **generic intrinsics must cover all the 10/10 numeric types**.

**Comment by fiigii** [comment]
@dotnet-bot test CROSS Check

File: src/mscorlib/src/System/Runtime/Intrinsics/X86/Avx2.PlatformNotSupported.cs [hunk 6]
```diff
@@ -722,7 +722,40 @@ public static class Avx2
         /// __m256i _mm256_inserti128_si256 (__m256i a, __m128i b, const int imm8)
         /// </summary>
         public static unsafe Vector256<ulong> Insert(Vector256<ulong> value, ulong* address, byte index) { throw new PlatformNotSupportedException(); }
-        
+
+        /// <summary>
+        /// __m256i _mm256_inserti128_si256 (__m256i a, __m128i b, const int imm8)
```
Comment (pentp) [inline src/mscorlib/src/System/Runtime/Intrinsics/X86/Avx2.PlatformNotSupported.cs]: These comments should be `__m256i _mm256_stream_load_si256 (__m256i const* mem_addr)`.

Comment (fiigii) [reply src/mscorlib/src/System/Runtime/Intrinsics/X86/Avx2.PlatformNotSupported.cs]: Ah, my bad. Let me fix these comments later. Thank you!
